### PR TITLE
openssh: switch to `openssl@3`

### PR DESCRIPTION
**Pull request by carlocab**
See #134251.

Split off from #134299 to unblock migration of `ldns`.

### CODE REVIEW

**Comment by carlocab** [comment]
Needs:
- #134299

**Comment by carlocab** [comment]
Failing on `master` too:
```
❯ lldb -- ./ssh-keygen -A
(lldb) target create "./ssh-keygen"
Current executable set to '/tmp/openssh-20230623-7195-4d1ep3/openssh-9.3p1/ssh-keygen' (x86_64).
(lldb) settings set -- target.run-args  "-A"
(lldb) r
Process 15308 launched: '/tmp/openssh-20230623-7195-4d1ep3/openssh-9.3p1/ssh-keygen' (x86_64)
Process 15308 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
    frame #0: 0x000000010000300e ssh-keygen`main(argc=0, argv=0x0000000000000000) at ssh-keygen.c:3355:32 [opt]
   3352         /* Ensure that fds 0, 1 and 2 are open or directed to /dev/null */
   3353         sanitise_stdfd();
   3354
-> 3355         __progname = ssh_get_progname(argv[0]);
   3356
   3357         seed_rng();
   3358
Target 0: (ssh-keygen) stopped.
warning: ssh-keygen was compiled with optimization - stepping may behave oddly; variables may not be available.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
  * frame #0: 0x000000010000300e ssh-keygen`main(argc=0, argv=0x0000000000000000) at ssh-keygen.c:3355:32 [opt]
    frame #1: 0x00007ff80f3fb41f dyld`start + 1903
```

File: Formula/openssh.rb [hunk 1]
```diff
@@ -67,6 +68,10 @@ def install
       # Ensure sandbox profile prefix is correct.
       # We introduce this issue with patching, it's not an upstream bug.
       inreplace "sandbox-darwin.c", "@PREFIX@/share/openssh", etc/"ssh"
+
+      # FIXME: `ssh-keygen` errors out when this is built with optimisation.
+      # TODO: Report this upstream.
+      ENV.O0 if Hardware::CPU.intel? && MacOS.version == :ventura && version == Version.new("9.3p1")
```
Comment (carlocab) [inline Formula/openssh.rb:74]: Left a `version` check here to try to make sure it's addressed at least at the next version bump.

**Comment by carlocab** [comment]
@Bo98 @fxcoudert Do you think [this](https://github.com/Homebrew/homebrew-core/pull/134721#issuecomment-1604385894) is a compiler bug? Only happening on Intel macOS 13. Using `ENV.O0` makes it go away.

**Review by p-linnane — APPROVED** [review]

**Comment by Bo98** [comment]
`argv` being null seems sound where `argc == 0`. Is `argc` always 0 or something?

**Comment by Bo98** [comment]
> `argv` being null is maybe(?) ok where `argc == 0` - will need to read specs to see.

Ah no it isn't:

> `argv[argc]` shall be a null pointer.

Seems like a compiler bug then.

**Pull request merged**